R/scanGenome.R

```diff
@@ -407,6 +407,133 @@ phasingScoreFourier <- function(se, gr, numCoef = 5) {
+#'                            c("6mA_1_10reads.bam", "6mA_2_10reads.bam"),
+#'                            package = "footprintR")
+#' rng <- GRanges("chr1", IRanges(6930000, 6940000))
+#' windowSize <- 2000
+#' windowStep <- 1000
+#' s <- seq(start(rng), end(rng) - windowSize + 1, by = windowStep)
+#' windowgr <- GRanges(seqnames = seqnames(rng),
+#'                     ranges = IRanges(start = s, width = windowSize))
+#' se <- readModBam(bamfiles = modbamfiles, regions = rng, level = "quickread",
+#'                  modbase = "a", trim = TRUE,
+#'                  BPPARAM = BiocParallel::SerialParam())
+#' seNRL <- estimateNRLwindows(se = se, gr = windowgr,
+#'                             BPPARAM = BiocParallel::SerialParam())
+#' assay(seNRL, "NRL")
+#' assay(seNRL, "NRL.CI95low")
+#' assay(seNRL, "NRL.CI95high")
+#'
+#' @importFrom SummarizedExperiment rowRanges colData assay assayNames ncol
+#' @importFrom GenomicRanges GRanges
+#' @importFrom IRanges findOverlaps
+#' @importFrom S4Vectors queryHits subjectHits
+#' @importFrom BiocParallel bplapply
+#'
+#' @export
+estimateNRLwindows <- function(se, gr,
+                               dmax = 1000L,
+                               assayName = "mod_prob",
+                               minModProb = 0.5,
+                               minDist = 140L,
+                               usePeaks = seq_len(5),
+                               span1 = 100/dmax,
+                               span2 = 1500/dmax,
+                               BPPARAM = BiocParallel::MulticoreParam(4L)) {
+    .assertVector(x = se, type = "RangedSummarizedExperiment")
+    .assertVector(x = gr, type = "GRanges")
+
+    if (nrow(se) > 0 && length(gr) > 0) {
+
+        ov <- findOverlaps(query = rowRanges(se), subject = gr)
+        indexL <- split(x = queryHits(ov),
+                        f = factor(subjectHits(ov), levels = seq.int(length(gr))),
+                        drop = FALSE)
+
+        # loop over samples
+        modprobL <- assay(se, assayName)
+        s <- start(se)
+        resL <- lapply(seq.int(ncol(se)), function(j) { # for each sample j
+            do.call(rbind, bplapply(indexL, function(i,  # for positions i in a window
+                                                     mymodprob = modprobL[[j]][i,],
+                                                     mys = s[i],
+                                                     myminModProb = minModProb,
+                                                     mydmax = dmax,
+                                                     myminDist = minDist,
+                                                     myusePeaks = usePeaks,
+                                                     myspan1 = span1,
+                                                     myspan2 = span2) {
+                if (length(i) > 0) { # nocov start
+                    cnt <- numeric(mydmax)
+                    for (r in seq.int(ncol(mymodprob))) {
```

Comment:
What do you think about replacing this for loop by something along the lines of 
```
tmp <- nnawhich(mymodprob, arr.ind = TRUE)[nnavals(mymodprob) >= myminModProb, ]
tmp[, 1] <- mys[tmp[, 1]]
tmp <- split(tmp[, 1], f = tmp[, 2])
for (pos in tmp) {
    calcAndCountDist(query = pos, reference = pos, cnt = cnt)
}
```
It seems to speed up things in my tests.

Reply:
That helped a lot - thank you! Now, most time (>60%) is spent in .estimateNRLfast, I think that needs some investigation now